### PR TITLE
Device provisioning API

### PR DESCRIPTION
**Pull request by Steve-Mcl**
## PR Strategy
```mermaid
gitGraph
    commit
    commit
    branch device-provisioning-api
    checkout device-provisioning-api
    commit
    commit
    branch device-provisioning-ui
    checkout device-provisioning-ui
    commit
    commit
    checkout device-provisioning-api
    merge device-provisioning-ui
    checkout main
    merge device-provisioning-api
    commit
```
Please review/approve but do not merge before #1664 

## Description

Adds endpoints (and supporting functions/logging/changes) for basic management of device auto provisioning AccessTokens.

* [Add new logger for capturing generic info](https://github.com/flowforge/flowforge/commit/ef6adc524378f331e658b2c82ca271a7cc1d3c27)
* [add device provisioning loggers](https://github.com/flowforge/flowforge/commit/c3c480cad3c012069d95bb0fed4857cdde937c05)
* [add DB migration to add id column to AccessToken](https://github.com/flowforge/flowforge/commit/18050dce81d75f2470e478b16dca988221a2fd8f)
* [add route permissions for provisioning APIs](https://github.com/flowforge/flowforge/commit/b4c6c92067148f6acb16d25190ad86e0e88fad4f)
* [add supporting functions for API methods/views](https://github.com/flowforge/flowforge/commit/37c8fe4ecaf489d13d181ccaba2eac4bedf796af)
* ~~[add platform setting device:auto-provisioning](https://github.com/flowforge/flowforge/commit/324615b7f52a8bcac9d2c8a0dd1ab19277afa85c)~~
   * Removed in [4687003](https://github.com/flowforge/flowforge/pull/1659/commits/46870034108cb7b71b7c7e5428ab2be869092860)
* [create API for management of provisioning tokens](https://github.com/flowforge/flowforge/commit/5a8a1e0ad03da02805ccbc9eb9df9c07625614c6)
* [add tests for provisioning](https://github.com/flowforge/flowforge/commit/22cb9e6d9ee2e33a462f60f8c57c54a9bf32b204)

## Related Issue(s)

Part of #1212 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [-] Changes `flowforge.yml`?
    - [-] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [-] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [x] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: forge/lib/templates.js [hunk 1]
```diff
@@ -19,7 +19,8 @@ module.exports = {
         'modules_denyList',
         'httpNodeAuth_type',
         'httpNodeAuth_user',
-        'httpNodeAuth_pass'
+        'httpNodeAuth_pass',
+        'deviceAutoProvision'
```
Comment (knolleary) [inline forge/lib/templates.js:23]: Is this a hangover from previous iterations? I don't see any relation to Templates for this feature.

Comment (Steve-Mcl) [reply forge/lib/templates.js]: Resolved in [591c018](https://github.com/flowforge/flowforge/pull/1659/commits/591c018563622472a1dadcf7ae17ee23ead73efa)

File: forge/routes/api/device.js [hunk 2]
```diff
@@ -70,13 +70,37 @@ module.exports = async function (app) {
...
+                // * If this is a Device Provisioning action: verify the device has the required scope
+                //   & session has been populated with provisioning data
+                // * If this is a User action: verify the user has the required role
+                if (request.session.provisioning) {
+                    // A device is auto-provisioning. First check the request body team matches the token
+                    // NOTE: If the token was not valid, the request would have been rejected by
+                    // the verifySession decorator & request.session.provisioning would not be populated
+                    const teamOK = request.body.team && request.body.team === request.session.provisioning.team
+                    if (teamOK) {
+                        request.deviceProvisioning = true
```
Comment (knolleary) [inline forge/routes/api/device.js:82]: Is this strictly needed if `request.session.provisioning` can also be used to know this is a valid provisioning token?

File: forge/db/migrations/20230205-01-add-id-to-access-tokens.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            // ensure that the data was copied correctly
+            const dataIntermediate = await context.select(null, 'AccessTokens2', { transaction: t })
+            if (dataOriginal.length !== dataIntermediate.length) {
+                throw new Error('Migration failed: data was not copied correctly')
+            }
+            const dataOriginalJSON = JSON.stringify(dataOriginal)
+            const dataIntermediateJSON = JSON.stringify(dataIntermediate.map((row) => { delete row.id; return row }))
+            if (dataOriginalJSON !== dataIntermediateJSON) {
+                throw new Error('Migration failed: data was not copied correctly')
+            }
```
Comment (knolleary) [inline forge/db/migrations/20230205-01-add-id-to-access-tokens.js:74]:
```suggestion
```

This breaks for me as it assumes the Object keys encode in an identical order. In my local testing the copy of data works fine, but the resulting JSON String is different.

I think this is unnecessary.

File: forge/db/migrations/20230205-01-add-id-to-access-tokens.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+            // ensure that the data was copied correctly
+            const dataFinal = await context.select(null, 'AccessTokens', { transaction: t })
+            if (dataOriginal.length !== dataFinal.length) {
+                throw new Error('Migration failed: data was not copied correctly')
+            }
+            const dataFinalJSON = JSON.stringify(dataFinal.map((row) => { delete row.id; return row }))
+            if (dataOriginalJSON !== dataFinalJSON) {
+                throw new Error('Migration failed: data was not copied correctly')
+            }
```
Comment (knolleary) [inline forge/db/migrations/20230205-01-add-id-to-access-tokens.js:89]:
```suggestion
```

As above - this assumes the key order is the same for the two objects.

File: forge/db/migrations/20230205-01-add-id-to-access-tokens.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                updatedAt: { type: DataTypes.DATE, allowNull: false }
+            }, { transaction: t })
+
+            // copy the data from the old table to the new table
+            await context.bulkInsert('AccessTokens2', dataOriginal, { transaction: t })
+
+            // ensure that the data was copied correctly
+            const dataIntermediate = await context.select(null, 'AccessTokens2', { transaction: t })
+            if (dataOriginal.length !== dataIntermediate.length) {
+                throw new Error('Migration failed: data was not copied correctly')
```
Comment (knolleary) [inline forge/db/migrations/20230205-01-add-id-to-access-tokens.js:68]: general comment - if a function can fail in multiple ways, don't give them all the same error message as it makes it impossible to identify which clause failed from just the error message.

**Review by knolleary — CHANGES_REQUESTED** [review]
Some changes needed on the migration

File: forge/db/migrations/20230205-01-add-id-to-access-tokens.js (hunk 5 above)
Comment (Steve-Mcl) [reply forge/db/migrations/20230205-01-add-id-to-access-tokens.js]: resolved in [8909bce](https://github.com/flowforge/flowforge/pull/1659/commits/8909bce4592f149ed9aa3b1d6e883eeb39e17701)

**Comment by Steve-Mcl** [comment]
@knolleary as discussed earlier, I have redone all manual operation testing using a local Postgres to ensure both starting from scratch and starting with existing DB (so that migration runs).

I have done this back and forth several times (after realising my dummy data was "expired") - but it was a good exercise in testing reliability. 

## Results...

### From empty / new PG instance
image


### From existing PG instance

#### Table before upgrade
image
#### Migration
image
#### DDL afterwards
image




image

File: forge/routes/api/device.js (hunk 2 above)
Comment (Steve-Mcl) [reply forge/routes/api/device.js]: Maybe a bit of belt + braces but ultimately a comfort blanket 😄 
Removed the explicit `request.deviceProvisioning` flag in [8ff3f7d](https://github.com/flowforge/flowforge/pull/1659/commits/8ff3f7df313252912df229bb10440425e612c70e)

**Review by knolleary — APPROVED** [review]
Once we have #1664 merged into this branch, and the tests running clean on this PR, I'm :+1: on merging.

**Pull request merged**